File: Queue_two_stack.swift
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this : Having doubt about writing time complexity


// Your code here along with comments explaining your approach
//Use two stack called stack and queueStack.
//insert the first push element into queueStack. then insert all other push elements into stack
//whenever pop move all the stack elements into queueStack  when queue stack is empty
//return queueStack last element for peek and pop



class MyQueue {

    var stack: [Int]!
    var queueStack: [Int]!
    
    /** Initialize your data structure here. */
    init() {
        stack = [Int]()
        queueStack = [Int]()
    }
    
    /** Push element x to the back of queue. */
    func push(_ x: Int) {
        if queueStack.isEmpty && stack.isEmpty {
            queueStack.append(x)
        } else {
            stack.append(x)
        }
        
    }
    
    /** Removes the element from in front of queue and returns that element. */
    func pop() -> Int {
        if queueStack.isEmpty && stack.isEmpty {
            return -1
        }
        moveStackElementsIntoQueueStack()
        return queueStack.removeLast()
    }
    
    /** Get the front element. */
    func peek() -> Int {
        if queueStack.isEmpty && stack.isEmpty {
            return -1
        }
        moveStackElementsIntoQueueStack()
        return queueStack.last!
    }
    
    /** Returns whether the queue is empty. */
    func empty() -> Bool {
        return queueStack.isEmpty && stack.isEmpty
    }
    
    /** put all stack elements into queueStack */
    func moveStackElementsIntoQueueStack() {
        if stack.isEmpty && queueStack.isEmpty {
            return
        }
         if queueStack.isEmpty {
             while !stack.isEmpty {
                 queueStack.append(stack.removeLast())
             }
          }
    }
}


/**
 * Your MyQueue object will be instantiated and called as such:
 * let obj = MyQueue()
 * obj.push(x)
 * let ret_2: Int = obj.pop()
 * let ret_3: Int = obj.peek()
 * let ret_4: Bool = obj.empty()
 */

File: Queue_two_stack.py
# // Time Complexity :  PUSH O(1), POP O(1) depending on input size though and intervals of pop, PEEK O(1), EMPTY O(1)
# // Space Complexity : for input of size n, O(N) but no extra memory used in functions
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : 
# // Your code here along with comments explaining your approach: initiated two stacks(lists), once pop aarives
# transfer for one time from s1 to s2, and keep popping from s2 and keep appending to s1.

class MyQueue:

    def __init__(self):
        self.s1=[]
        self.s2=[]

    def push(self, x: int) -> None:
        self.s1.append(x)

    def pop(self) -> int:
        if(len(self.s2)==0):
            i=len(self.s1)
            while(i!=0):
                self.s2.append(self.s1.pop())
                i-=1
        return self.s2.pop()

    def peek(self) -> int:
        if(len(self.s2)!=0):
            return self.s2[-1]
        return self.s1[0]

    def empty(self) -> bool:
        if(len(self.s1)==0 and len(self.s2)==0):
            return True
        return False

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()